### ISSUE TITLE
[Refactor] Retrieve correct CLIENT_ID from backend

### ISSUE THREAD

**Issue by rinukkusu** [taniarascia/takenote#266]
**Problem**
In the README.md file on setting up the takenote application it says:
> Finally, in `LandingPage.tsx`, replace the `clientId` variable with your app's client id.

This is really coming back to my other issue regarding upgradability with the `latest` tag. Now I also have to edit a file in the docker container itself? I'm sure we can do better! 😄 

**Solution**
Retrieve the correct `clientId` from the server itself instead of hardcoding it, since we already set the environment variables for the server so we already have the correct `clientId` in one place.

Ideas? Objections?

**Comment by taniarascia** [comment]
You’re right, it was a temporary hack just to get auth up and running. I’ll look into a better method of getting the variable into React soon.

**Comment by taniarascia** [comment]
It took quite a bit of messing around to get both Node and Webpack to share the same env config while still allowing the deployment to work, but it’s all set now!

**Comment by rinukkusu** [comment]
Does not look like it's fixed yet :crying_cat_face: 
Still get your clientId in the updated docker image, when I try to sign in.

**Comment by taniarascia** [comment]
@rinukkusu Ah...I thought you meant for local development. I had to bypass the production one because reading from `dotenv` in Webpack REQUIRES an `.env` file, but I only pass variables via command line for Docker in the one-liner deployment. I tried creating an `.env` with client id, but it wasn't working.

**Comment by taniarascia** [comment]
Okay, I figured it out. I can be sure: 

- `.env` exists in local development mode
- `.env` exists in local production mode, and 
- `.env` does not exist in Docker or deployment

So just pass it as a build arg for Docker, and the rest will pull from .env.

```js
const getEnvFromDotEnvFile = dotenv.config()
let envKeys
if (getEnvFromDotEnvFile.error) {
  console.log('Getting environment variables from build args for production') // eslint-disable-line
  envKeys = {
    'process.env.CLIENT_ID': JSON.stringify(process.env.CLIENT_ID),
    'process.env.NODE_ENV': JSON.stringify('production'),
  }
} else {
  envKeys = { 'process.env.CLIENT_ID': JSON.stringify(getEnvFromDotEnvFile.parsed['CLIENT_ID']) }
}
```

The only issue is that Docker still won't work locally due to refusing to set the `secure` cookie, if you're running on `localhost`.

```js
export const thirtyDayCookie = {
  maxAge: 60 * 60 * 1000 * 24 * 30,
  secure: isProduction,// <-- this will be set to true in Docker
  httpOnly: true,
  sameSite: true,
}
```

I was also able to fix this issue by removing the `NODE_ENV` from the start command, and only having it on the build command, so you can set the Express server to production, thus allowing non-https cookies locally.

```js
docker build --build-arg CLIENT_ID=xxx -t takenote:dev .
docker run -p CLIENT_ID=xxx CLIENT_SECRET=xxxx NODE_ENV=development 5000:5000 takenote:dev
```

Now you can run Docker locally without updating any code. 

Once I merge the PR in, this will be ready. Thanks for opening this issue, I learned a lot about getting it all up and running!

**Comment by taniarascia** [comment]
The issue here becomes if you try to use the Docker Hub build, it's already been built with the client id baked in. There is no way around this. 

The Docker image MUST be built with the client ID because it is not possible to dynamically retrieve environment variables in a production Webpack build. Unlike Node, which is actively running, the React front end is static. It must be baked into the build.

So being able to build a local Docker image with your own variables is the best I can do.

**Comment by taniarascia** [comment]
Closed in #268. Let me know if you have any further issues!
